src/builder.rs

```diff
@@ -254,93 +217,347 @@ impl<'a> RtpPacketBuilder<'a> {
             }
         }
 
-        Ok(())
+        self.calculate_size()
     }
 
-    /// Write this packet into `buf`.  On success returns the number of bytes written or an
-    /// `RtpWriteError` on failure.
-    pub fn write_into(self, buf: &mut [u8]) -> Result<usize, RtpWriteError> {
-        self.write_preconditions()?;
+    /// Write this packet using writer without any validity checks
+    pub fn write_unchecked<O>(
+        &self,
+        writer: &mut dyn RtpPacketWriter<Payload = P, Extension = E, Output = O>,
+    ) -> O {
+        let mut hdr = [0; RtpPacket::MIN_RTP_PACKET_LEN];
+        self.write_header_into(&mut hdr);
+        writer.push(hdr.as_ref());
+
+        for csrc in self.csrcs.iter() {
+            writer.push(csrc.to_be_bytes().as_ref());
+        }
+
+        if let Some((ext_id, ext_data)) = self.extension.as_ref() {
+            writer.push(ext_id.to_be_bytes().as_ref());
+            writer.push(((ext_data.len() / 4) as u16).to_be_bytes().as_ref());
+            writer.push_extension(ext_data);
+        }
+
+        for payload in self.payloads.iter() {
+            writer.push_payload(payload);
+        }
+
+        if let Some(padding) = self.padding {
+            writer.padding(padding);
+        }
 
-        let size = self.calculate_size()?;
-        if size > buf.len() {
-            return Err(RtpWriteError::OutputTooSmall(size));
+        writer.finish()
+    }
+
+    /// Write the packet using writer.
+    pub fn write<O>(
+        &self,
+        writer: &mut dyn RtpPacketWriter<Payload = P, Extension = E, Output = O>,
```

Comment:
Why `dyn` instead of `impl`?

Reply:
No reason :)

Reply:
If you have `impl` then you can also use `dyn` in it, the other way around not :) Leaves the choice to the caller.